interface/ceed-operator.c

```diff
@@ -1264,22 +1275,525 @@ int CeedOperatorLinearAssemblePointBlockDiagonal(CeedOperator op,
 int CeedOperatorLinearAssembleAddPointBlockDiagonal(CeedOperator op,
     CeedVector assembled, CeedRequest *request) {
   int ierr;
-  Ceed ceed = op->ceed;
-  ierr = CeedOperatorCheckReady(ceed, op); CeedChk(ierr);
+  ierr = CeedOperatorCheckReady(op); CeedChk(ierr);
 
   // Use backend version, if available
   if (op->LinearAssembleAddPointBlockDiagonal) {
-    ierr = op->LinearAssembleAddPointBlockDiagonal(op, assembled, request);
-    CeedChk(ierr);
+    return op->LinearAssembleAddPointBlockDiagonal(op, assembled, request);
   } else {
     // Fallback to reference Ceed
     if (!op->opfallback) {
       ierr = CeedOperatorCreateFallback(op); CeedChk(ierr);
     }
     // Assemble
-    ierr = op->opfallback->LinearAssembleAddPointBlockDiagonal(op->opfallback,
-           assembled, request); CeedChk(ierr);
+    return CeedOperatorLinearAssembleAddPointBlockDiagonal(op->opfallback,
+           assembled, request);
+  }
+}
+
+/**
+   @brief Build nonzero pattern for non-composite operator.
+
+   Users should generally use CeedOperatorLinearAssembleSymbolic()
+
+   @ref Developer
+**/
+int CeedSingleOperatorAssembleSymbolic(CeedOperator op, CeedInt offset,
```

Comment:
I think it would help if we clearly indicate in the function name if we are summing into a target vector or overwriting. This will help when I refactor the diagonal assembly to have an interface level impl with some reduced code duplication.

See the `/cpu/self/ref` impl of diagonal assembly for context:

<details><summary>Click to Show/Hide</summary>

```
//------------------------------------------------------------------------------
// Assemble composite diagonal common code
//------------------------------------------------------------------------------
static inline int CeedOperatorLinearAssembleAddDiagonalCompositeCore_Ref(
  CeedOperator op, CeedVector assembled, CeedRequest *request,
  const bool pointBlock) {
  int ierr;
  CeedInt numSub;
  CeedOperator *subOperators;
  ierr = CeedOperatorGetNumSub(op, &numSub); CeedChkBackend(ierr);
  ierr = CeedOperatorGetSubList(op, &subOperators); CeedChkBackend(ierr);
  for (CeedInt i = 0; i < numSub; i++) {
    ierr = CeedOperatorAssembleAddDiagonalCore_Ref(subOperators[i], assembled,
           request, pointBlock); CeedChkBackend(ierr);
  }
  return CEED_ERROR_SUCCESS;
}

//------------------------------------------------------------------------------
// Assemble Linear Diagonal
//------------------------------------------------------------------------------
static int CeedOperatorLinearAssembleAddDiagonal_Ref(CeedOperator op,
    CeedVector assembled, CeedRequest *request) {
  int ierr;
  bool isComposite;
  ierr = CeedOperatorIsComposite(op, &isComposite); CeedChkBackend(ierr);
  if (isComposite) {
    return CeedOperatorLinearAssembleAddDiagonalCompositeCore_Ref(op, assembled,
           request, false);
  } else {
    return CeedOperatorAssembleAddDiagonalCore_Ref(op, assembled, request, false);
  }
}

//------------------------------------------------------------------------------
// Assemble Linear Point Block Diagonal
//------------------------------------------------------------------------------
static int CeedOperatorLinearAssembleAddPointBlockDiagonal_Ref(CeedOperator op,
    CeedVector assembled, CeedRequest *request) {
  int ierr;
  bool isComposite;
  ierr = CeedOperatorIsComposite(op, &isComposite); CeedChkBackend(ierr);
  if (isComposite) {
    return CeedOperatorLinearAssembleAddDiagonalCompositeCore_Ref(op, assembled,
           request, true);
  } else {
    return CeedOperatorAssembleAddDiagonalCore_Ref(op, assembled, request, true);
  }
}
```

<\details>

Reply:
We are kind of "appending" to the target vector, so for a composite operator you get entries in the `values` vector for all the nonzero entries in each of the sub-operators. I'll try to update the documentation to make this clearer - are there names for the functions that would also make this clear?

Reply:
With diagonal assembly, we use `Assemble` for functions that overwrite the target vector and `AssembleAdd` for functions that sum into the target vector. I set it up this way to support MFEM's ability to compose diagonal assembly for multiple integrators implemented in libCEED.

I think we should
1) add `Add` to the name of helper functions that sum into their vectors
2) make `Apply` and `ApplyAdd` versions of full assembly function

Reply:
I understand that. This is a different case from diagonal assembly and I don't think those conventions apply here. `Add` variants do not make sense here because there is no way to ensure that you are adding into the same nonzero structure, so what we do is *append* the values to a coordinate-format-like data structure, and then the client code (eg MFEM) can assemble or add or do whatever it wants.

Reply:
So down the line MFEM will be obligated to add together the separately assembled mass and diffusion operators? I'm fine with making that MFEM's problem rather than libCEED's.

I still think it would be valuable to indicate in the name when you overwrite vs *append*. If you prefer the word *append*, go for it. I'll figure out how that fit with the future refactoring of assembly/diagonal assembly at the interface level when I get there.